### PR TITLE
Fix `call { true or next }` producing invalid js

### PR DESCRIPTION
**Pull request by elia**
Fixes https://github.com/opal/opal/issues/810

I'm not too happy with this solution in terms of produced code, on the other hand maybe some `if` can be simplified into ternary or `&&`/`||`.

I'll try to retain the simpler form when not needed before putting this into "ready for review".

### CODE REVIEW

**Comment by hmdne** [comment]
I would suggest rewriting `and/or` as `if` only if the rhs is one of the `next/return/break` OR rhs.rhs is one of those, ad infinitum.

**Comment by elia** [comment]
@hmdne I thought about it and did some performance testing, and after all I think it's better to converge both and/or and if to a single optimized implementation. If we're able to use a ternary operator whenever it makes sense for both of them is a double win, moreover we'll need to support in logical operators everything that is available for if, in terms of _returnability_ and other features.

@iliabylich you have any considerations on this or reasons not to do this?

File: lib/opal/rewriters/returnable_logic.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+module Opal
+  module Rewriters
+    class ReturnableLogic < Base
+      def next_tmp
+        @counter ||= 0
+        @counter += 1
+        "$ret_or_#{@counter}"
+      end
+
+      def reset_tmp_counter!
```
Comment (iliabylich) [inline lib/opal/rewriters/returnable_logic.rb:14]: Does anyone use this method? It was a part of another rewriter to simplify testing (i.e. to have consistent `tmp_name_1` in tests)

File: lib/opal/rewriters/returnable_logic.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        @counter = nil
+      end
+
+      def on_or(node)
+        lhs, rhs = *node.children
+        lhs_tmp = next_tmp
+
+        node.updated(:if, [s(:lvasgn, lhs_tmp, process(lhs)), s(:js_tmp, lhs_tmp), process(rhs)])
```
Comment (iliabylich) [inline lib/opal/rewriters/returnable_logic.rb:22]: oh I have no idea why Opal got it now. Looks like are great idea to me, and I believe we could implement it much, much earlier.

IIRC sorbet has a very similar idea of "ast desugaring". Basically they rewrite AST to have as few nodes as possible by converting them into other "trivial" nodes that they support by default. For example, they rewrite `case` into a chain of `if`s

**Review by iliabylich — APPROVED** [review]
LGTM overall, sorry I should've included both comments into a review, but I missclicked 😄

File: lib/opal/rewriters/returnable_logic.rb (hunk 1 above)
Comment (elia) [reply lib/opal/rewriters/returnable_logic.rb:14]: Yeah, that's why I added it too initially, then I changed the tests in a way that doesn't require it, some cleanup is due there, maybe move the helper up the inheritance chain, I'll open an issue as a reminder 👍

File: lib/opal/rewriters/returnable_logic.rb (hunk 2 above)
Comment (elia) [reply lib/opal/rewriters/returnable_logic.rb:22]: This is great! https://github.com/sorbet/sorbet/blob/master/ast/desugar/Desugar.cc
I'll see if I can back port other stuff from it, and get less and more robust code 👍

**Pull request merged**